data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriters.java

```diff
@@ -231,7 +231,23 @@ public void nonNullWrite(int rowId, String data, ColumnVector output) {
 
     @Override
     public void nonNullWrite(int rowId, ByteBuffer data, ColumnVector output) {
-      ((BytesColumnVector) output).setRef(rowId, data.array(), 0, data.array().length);
+      // When there is a backing heap based byte array, we avoid the overhead of
+      // copying it.
+      if (data.hasArray()) {
+        // Don't assume the we're reading in the entire backing array.
+        // Using slice as any mutations to the data at rowId of output
+        // would also be visible in `data`.
+        ByteBuffer slice = data.slice();
+        ((BytesColumnVector) output).setRef(rowId, slice.array(), 0, slice.array().length);
```

Comment:
I updated this line to not assume that the entire backing array of `data` is what we're reading in. Via the use of slice, a shallow copy takes place (which mimics the previous shallow copy). Does this resolve your previous concern?

I can clean up the comments etc once I know if this resolves any on going issues or not @rdblue . Thanks so much for your thorough review!

Reply:
Actually, my solution with slice causes a ByteBufferBackingArray warning, which could be suppressed as we called `.hasArray` before. But at this point suppression seems silly given the work that went in to correcting this.

If anybody has any solutions which doesn't involve a deep copy, I'm open to suggestions.

Reply:
This isn't a correct use of `slice`. Slice is like `duplicate`, except that it guarantees that `position` will be 0 and `capacity` is equal to `limit`. It still uses the same backing bytes and will not copy, which we would not want even if it did. Because this still uses the same backing array, we still need to use `arrayOffset`.

There isn't any benefit to using `slice` here since accessing the backing byte array means that we don't modify the original buffer.

Reply:
I've factored out a `duplicate` and then reused that in both portions, on heap and off, and changed the `position` to be `arrayOffset` when on heap. Thanks for your review.